subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/MultipleCandidateMatcher.java

```diff
@@ -329,10 +331,25 @@ private void removeCandidatesWithValueNotIn(int a, Set<Object> matchedValues) {
     }
 
     private void disambiguateWithExtraAttributes() {
```

Comment:
It's not at all obvious how these `disambiguate` methods work, since they cause effects not by returning a value, but by the side-effect of halting iteration through the attributes via `return`-ing.  I know there is a class-level javadoc of the process, but I think each method is also complex enough to deserve their own, and that they would work better than the inline comments, especially after a significant change like this.

Reply:
As with the other comment, I would prefer to revisit this in a follow-up PR, though I agree better documentation is needed here